src/preprocessing/passes/unconstrained_simplifier.cpp

```diff
@@ -858,9 +858,14 @@ PreprocessingPassResult UnconstrainedSimplifier::applyInternal(
     for (size_t i = 0, asize = assertions.size(); i < asize; ++i)
     {
       Node a = assertions[i];
-      Node as = rewrite(d_substitutions.apply(a));
-      // replace the assertion
-      assertionsToPreprocess->replace(i, as);
+      Node as = d_substitutions.apply(a);
+      if (as != a)
```

Comment:
this is again repeating a logic that is already in `replace`. I assume you added it because of the `ensureRewritten` that you have below, but I did not quite understood what it is for, from the description of the PR. Could you clarify?

Reply:
But why is it needed now when it wasn't before? In the PR description you say that the new method is to "ensure their trust steps are modulo rewriting", which I don't quite understand.

Reply:
The old code was worse in 2 ways: 
(1) it was not as efficient as it could be since it called rewrite even if `d_substitutions.apply` had no effect.
(2) it added `t = rewrite(apply(t))` as a trust step. The new functionality allows us to have a trust step `t = apply(t)`, where `apply(t)=rewrite(apply(t))` is handled by a rewrite proof generator.

Reply:
Ah, I see now.